src/Effect/AVar.purs

```diff
@@ -50,7 +63,51 @@ kill err avar = Fn.runFn3 _killVar ffiUtil err avar
 -- | the AVar becomes available. Returns an effect which will remove the
 -- | callback from the pending queue.
 put ∷ ∀ a. a → AVar a → AVarCallback Unit → Effect (Effect Unit)
-put value avar cb = Fn.runFn4 _putVar ffiUtil value avar cb
+put = windowPut (const PushTail)
+
+-- | Puts a value into an AVar using a strategy determined
+-- | dynamically on the basis of the number of pending puts
+-- | (i.e. not including the current value of the AVar, if non-EMPTY).
+windowPut
+  :: ∀ a
+   . (Int -> Operation)
+  -> a
+  -> AVar a
+  -> AVarCallback Unit
+  -> Effect (Effect Unit)
+windowPut strategy value avar cb =
+  case strategy (_pendingPuts avar) of
+    -- Do nothing with the queue
+    Ignore -> pure $ pure unit
+    -- Propagate an exception to the callback
+    Fail e -> do
+      cb $ Left e
+      pure $ pure unit
+    -- Kill the internal queue and propagate the exception
+    Halt e -> do
+      kill e avar
+      cb $ Left e
+      pure $ pure unit
+    -- Push callback onto the head
+    PushHead -> Fn.runFn4 _consVar ffiUtil value avar cb
+    -- Push callback onto the tail
+    PushTail -> Fn.runFn4 _snocVar ffiUtil value avar cb
+    -- Drop the head, and push onto the tail
+    DropHead e -> do
+      void $ Fn.runFn3 _tailPuts ffiUtil e avar
+      Fn.runFn4 _snocVar ffiUtil value avar cb
```

Comment:
Should these operations be atomic? Right now tailPuts/initPuts invokes a callback, which may end up doing something that affects the internal AVar state.

Reply:
yes, perhaps there is a need to create atomic versions of these combinations where the callbacks corresponding to dropped entries are invoked after the avar has been updated. I'll submit an update.

It is still possible for the callbacks to invalidate the property being enforced by the strategy. Strategies can still be coded to recover from such deviations.